### ISSUE TITLE
Cancel running effect when no longer needed?

### ISSUE THREAD

**Issue by jooooel** [ngrx/effects#133]
I'm using `ngrx/Effects` to load items from an API (I also have a reducer that updates the store on the different actions). 

If the user navigates away from `main-view` to another view before the items have loaded, how can I cancel that running request?

**main-view.component.ts**
```typescript
ngOnInit() {
    this.store.dispatch({ type: LOAD_ITEMS });
    ...
}
```

**item.effects.ts**
```typescript
@Effect() loadItems = this.actions
        .ofType(LOAD_ITEMS)
        .switchMap(() =>
            this.http.get(url)

                // If successful, dispatch success action with result
                .map((data: any) => ({ type: LOAD_ITEMS_SUCCESS, payload: data })

                // If request fails, dispatch failed action
                .catch(() => Observable.of({ type: LOAD_ITEMS_ERROR, payload: "Failed to load items" }))
        );
```

**Comment by littleStudent** [comment]
you can only cancel that request if you have access to the subscription and you can call `unsubscribe()`.
You could create your own `http-service` that keeps track of all running requests. And then you can dispatch another action `CANCEL_REQUEST` where you call `unsubscribe` in your custom http service.
Does this make sense?

**Comment by fxck** [comment]
takeUntil route change event?

**Comment by jooooel** [comment]
@littleStudent I was hoping that there was something built in to `ngrx/Effects` that would let me cancel currently running effects or something. I was looking at `EffectsSubscription` first, but I had misunderstood what it was for.

**Comment by jooooel** [comment]
@fxck Hm, maybe that something like that would work yes... 

I thought this would be a pretty common case that most people would need a fix for, maybe it isn't?

**Comment by kylecordes** [comment]
Yes, this can be handled easily and nicely within Effects. I don't have a code snippet with me as I write this. You certainly do not need to manually call unsubscribe or anything like that. Rather, like so:

* Watch both the LOAD__ITEMS and whatever action means the user is no longer on the screen where the data matters.
* When you do so you will get a single stream of both kinds of actions, the syntax is shown in the docs `actions$.ofType('LOGIN', 'LOGOUT');`.
* Inside of the `.switchMap(`, do an ordinary if, and return an `Observable.of(whatever)` instead of the HTTP Observable for the navigated-away case.
* That's it -  as the user navigates away, the switchMap will take care of canceling the now-moot Http observable in-flight for you

**Comment by mflorence99** [comment]
Thanks @kylecordes! I know the code is fairly obvious, but after reading your very helpful post I thought that for completeness I'd add the code I got working for a Web socket (where the unsubscribe is more important). This thread was just what I needed to know.
```
  @Effect() listen: Observable<Action> = this.actions
    .ofType(sigmasActions.ActionTypes.LISTEN, sigmasActions.ActionTypes.UNLISTEN)
    .startWith(new sigmasActions.ListenAction())
    .switchMap((action: Action) => {
      if (action.type === sigmasActions.ActionTypes.UNLISTEN)
        return of();
      else return this.ledService.getSigmas()
        .map((payload: Sigma[]) => new sigmasActions.ListenSuccessAction(payload))
        .catch((error: Response) => of(new sigmasActions.ListenFailureAction(handleError(error))));
    });
```

**Comment by PerfectPixel** [comment]
@MikeRyanDev you added the fixed in 4.0 label. Could you explain how the API changed in 4.0? I did not find anything on the platform repo.

(I was looking to cancel all running effects in case the user logs out, then I came across this issue)
